src/generate/utils/State_AI.cpp

```diff
@@ -48,4 +52,123 @@ namespace spy::gameplay {
 
         return leafSuccessors;
     }
+
+    void State_AI::addDamage(const spy::character::Character &character, int damage) {
+        if (character.hasProperty(character::PropertyEnum::TOUGHNESS)) {
+            damage /= 2;
+        }
+
+        auto id = character.getCharacterId();
+        if (hpDiff.find(id) == hpDiff.end()) {
+            // entry does not exist
+            hpDiff[id] = -damage;
+        } else {
+            // entry already exists
+            hpDiff[id] -= damage;
+        }
+
+    }
+
+    void State_AI::modStateChance(const character::Character &character, double successChance) {
+        using spy::character::PropertyEnum;
+        using spy::gadget::GadgetEnum;
+
+        // character with clammy clothes only has half the chance of success
+        if (character.hasProperty(PropertyEnum::CLAMMY_CLOTHES) ||
+            character.hasProperty(PropertyEnum::CONSTANT_CLAMMY_CLOTHES)) {
+            successChance /= 2;
+        }
+
+        // if char has tradecraft and not mole die, prob. test is repeated
+        if (character.hasProperty(PropertyEnum::TRADECRAFT)
+            && !character.hasGadget(GadgetEnum::MOLEDIE)) {
+            successChance = 1 - std::pow(1 - successChance, 2);
+        }
+
+        stateChance *= successChance;
+    }
+
+    std::pair<std::vector<State_AI>, bool> State_AI::handleHoneyTrap(const GadgetAction &op, const MatchConfig &config,
+                                                                     const libclient::LibClient &libClient) {
+        std::vector<State_AI> honeyTrapStates;
+        bool foundAlternatives = false;
+
+        auto sourceChar = this->getCharacters().findByUUID(op.getCharacterId());
+        auto targetChar = util::GameLogicUtils::findInCharacterSetByCoordinates(this->getCharacters(),
+                                                                                op.getTarget());
+
+        if (!this->wasHoneyTrapUsed && config.getHoneyTrapSuccessChance() != 0 &&
+            targetChar != this->getCharacters().end()) {
+            if (targetChar->hasProperty(spy::character::PropertyEnum::HONEY_TRAP)) {
+                auto otherOps = this->getHoneyTrapAlternatives(op, config);
+                if (!otherOps.empty()) {
+                    foundAlternatives = true;
+
+                    // copy and modify state
+                    State_AI s = *this;
+                    s.wasHoneyTrapUsed = true;
+                    s.modStateChance(*sourceChar,
+                                     config.getHoneyTrapSuccessChance() / static_cast<double>(otherOps.size()));
+
+                    // reset stuff from executeGadget base method
+                    s.operationsLeadingToState.pop_back();
+                    s.usedGadgets.pop_back();
+                    auto character = s.getCharacters().getByUUID(op.getCharacterId());
+                    character->setActionPoints(character->getActionPoints() + 1);
+                    s.isLeafState = false;
```

Comment:
warum wird das hier gemacht?

Reply:
Das sind die Sachen, die ich in der allgemeinen executeGaget(...) mache, damit ich das nicht in jeder einzelnen Gadget Methode machen muss. Da ich aber bei Honey Trap nicht die spezielle Gadget Methode aufrufen kann muss ich das "zurücksetzen" damit das nicht doppelt drin ist.

Reply:
Mir ist aber gerade aufgefallen, dass ich das in der allgemeinen executeGadget(...) über die `wasHoneyTrapUsed` Variable im state unterbinden kann. Ist jetzt geändert

Reply:
ok, also wird  hier quasi die ursprüngliche gadgetaction ersetzt?